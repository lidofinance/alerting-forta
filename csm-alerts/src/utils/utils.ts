import { Contract, EventFilter, Event } from 'ethers'

const LOG_FILTER_CHUNK = 2000

export async function getLogsByChunks(contract: Contract, filter: EventFilter, startblock: number, endBlock: number) {
  const events: Event[] = []
  let endBlockChunk
  let startBlockChunk = startblock
  do {
    endBlockChunk =
      endBlock > startBlockChunk + LOG_FILTER_CHUNK - 1 ? startBlockChunk + LOG_FILTER_CHUNK - 1 : endBlock
    const eventsChunk = await contract.queryFilter(filter, startBlockChunk, endBlockChunk)
    events.push(...eventsChunk)
    startBlockChunk = endBlockChunk + 1
  } while (endBlockChunk < endBlock)
  return events
}
