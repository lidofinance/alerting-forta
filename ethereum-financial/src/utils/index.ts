import BigNumber from "bignumber.js";

const TEN_TO_18 = new BigNumber(10).pow(18);

export function formatEth(amount: any, dp: number): string {
  return new BigNumber(String(amount)).div(TEN_TO_18).toFixed(dp);
}

export function formatDelay(fullDelaySec: number) {
  let sign = fullDelaySec >= 0 ? 1 : -1;
  let delayHours = 0;
  let delayMin = Math.floor((sign * fullDelaySec) / 60);
  let delaySec = sign * fullDelaySec - delayMin * 60;
  if (delayMin >= 60) {
    delayHours = Math.floor(delayMin / 60);
    delayMin -= delayHours * 60;
  }
  return (
    (sign == 1 ? "" : "-") +
    (delayHours > 0 ? `${delayHours} hrs ` : "") +
    (delayMin > 0 ? `${delayMin} min ` : "") +
    `${delaySec} sec`
  );
}
