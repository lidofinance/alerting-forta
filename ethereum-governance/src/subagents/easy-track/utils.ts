export const getMotionType = (
  types: Map<string, string>,
  evmScriptFactory: string,
) => {
  return types.get(evmScriptFactory.toLowerCase()) || "New ";
};

export const getMotionLink = (motionId: string) => {
  return `[${motionId}](https://easytrack.lido.fi/motions/${motionId})`;
};
