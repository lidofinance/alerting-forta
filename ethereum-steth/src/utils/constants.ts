import BigNumber from 'bignumber.js'

export type ERC20 = {
  decimals: number
  name: string
}

export type Address = {
  DEPOSIT_SECURITY_ADDRESS: string
  BURNER_ADDRESS: string
  LDO_ADDRESS: string
  LIDO_STETH_ADDRESS: string
  WSTETH_ADDRESS: string
  WITHDRAWALS_QUEUE_ADDRESS: string
  DEPOSIT_EXECUTOR_ADDRESS: string
  INSURANCE_FUND_ADDRESS: string
  DAI_ADDRESS: string
  USDT_ADDRESS: string
  USDC_ADDRESS: string
  GATE_SEAL_DEFAULT_ADDRESS: string
  EXIT_BUS_ORACLE_ADDRESS: string
  GATE_SEAL_FACTORY_ADDRESS: string
  WITHDRAWALS_VAULT_ADDRESS: string
  EL_REWARDS_VAULT_ADDRESS: string
  KNOWN_ERC20: Map<string, ERC20>
}

export const ETH_DECIMALS = new BigNumber(10).pow(18)

const DEPOSIT_SECURITY_ADDRESS: string = '0xc77f8768774e1c9244beed705c4354f2113cfc09'
const BURNER_ADDRESS: string = '0xd15a672319cf0352560ee76d9e89eab0889046d3'
const LDO_ADDRESS: string = '0x5a98fcbea516cf06857215779fd812ca3bef1b32'
const LIDO_STETH_ADDRESS: string = '0xae7ab96520de3a18e5e111b5eaab095312d7fe84'
const WSTETH_ADDRESS: string = '0x7f39c581f595b53c5cb19bd0b3f8da6c935e2ca0'
const WITHDRAWALS_QUEUE_ADDRESS: string = '0x889edc2edab5f40e902b864ad4d7ade8e412f9b1'
const DEPOSIT_EXECUTOR_ADDRESS: string = '0xf82ac5937a20dc862f9bc0668779031e06000f17'
const INSURANCE_FUND_ADDRESS: string = '0x8b3f33234abd88493c0cd28de33d583b70bede35'
const DAI_ADDRESS: string = '0x6b175474e89094c44da98b954eedeac495271d0f'
const USDT_ADDRESS: string = '0xdac17f958d2ee523a2206206994597c13d831ec7'
const USDC_ADDRESS: string = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'
const GATE_SEAL_DEFAULT_ADDRESS: string = '0x1ad5cb2955940f998081c1ef5f5f00875431aa90'
const EXITBUS_ORACLE_ADDRESS: string = '0x0de4ea0184c2ad0baca7183356aea5b8d5bf5c6e'
const GATE_SEAL_FACTORY_ADDRESS: string = '0x6c82877cac5a7a739f16ca0a89c0a328b8764a24'
const WITHDRAWALS_VAULT_ADDRESS: string = '0xb9d7934878b5fb9610b3fe8a5e441e8fad7e293f'
const EL_REWARDS_VAULT_ADDRESS: string = '0x388c818ca8b9251b393131c08a736a67ccb19297'

const KNOWN_ERC20 = new Map<string, ERC20>([
  [LIDO_STETH_ADDRESS, { decimals: 18, name: 'stETH' }],
  [WSTETH_ADDRESS, { decimals: 18, name: 'wstETH' }],
  [LDO_ADDRESS, { decimals: 18, name: 'LDO' }],
  [DAI_ADDRESS, { decimals: 18, name: 'DAI' }],
  [USDT_ADDRESS, { decimals: 6, name: 'USDT' }],
  [USDC_ADDRESS, { decimals: 6, name: 'USDC' }],
])

export const Address: Address = {
  DEPOSIT_SECURITY_ADDRESS: DEPOSIT_SECURITY_ADDRESS,
  BURNER_ADDRESS: BURNER_ADDRESS,
  LDO_ADDRESS: LDO_ADDRESS,
  LIDO_STETH_ADDRESS: LIDO_STETH_ADDRESS,
  WSTETH_ADDRESS: WSTETH_ADDRESS,
  WITHDRAWALS_QUEUE_ADDRESS: WITHDRAWALS_QUEUE_ADDRESS,
  DEPOSIT_EXECUTOR_ADDRESS: DEPOSIT_EXECUTOR_ADDRESS,
  INSURANCE_FUND_ADDRESS: INSURANCE_FUND_ADDRESS,
  DAI_ADDRESS: DAI_ADDRESS,
  USDT_ADDRESS: USDT_ADDRESS,
  USDC_ADDRESS: USDC_ADDRESS,
  GATE_SEAL_DEFAULT_ADDRESS: GATE_SEAL_DEFAULT_ADDRESS,
  EXIT_BUS_ORACLE_ADDRESS: EXITBUS_ORACLE_ADDRESS,
  GATE_SEAL_FACTORY_ADDRESS: GATE_SEAL_FACTORY_ADDRESS,
  WITHDRAWALS_VAULT_ADDRESS: WITHDRAWALS_VAULT_ADDRESS,
  EL_REWARDS_VAULT_ADDRESS: EL_REWARDS_VAULT_ADDRESS,
  KNOWN_ERC20: KNOWN_ERC20,
}
