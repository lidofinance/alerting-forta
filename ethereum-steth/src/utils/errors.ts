import { Finding } from '../generated/proto/alert_pb'

export const NetworkErrorFinding = 'NETWORK-ERROR'
export const BotOutdatedAlertID = 'L1-BLOCK-OUTDATED'

export function networkAlert(err: Error, name: string, desc: string): Finding {
  const f: Finding = new Finding()
  f.setName(name)
  f.setDescription(desc)
  f.setAlertid(NetworkErrorFinding)
  f.setSeverity(Finding.Severity.UNKNOWN)
  f.setType(Finding.FindingType.DEGRADED)
  f.setProtocol('ethereum')

  const m = f.getMetadataMap()
  m.set('stack', `${err.stack}`)
  m.set('message', `${err.message}`)
  m.set('name', `${err.name}`)

  return f
}

export function dbAlert(err: Error, name: string, desc: string): Finding {
  const f: Finding = new Finding()
  f.setName(name)
  f.setDescription(desc)
  f.setAlertid('DB-ERROR')
  f.setSeverity(Finding.Severity.UNKNOWN)
  f.setType(Finding.FindingType.DEGRADED)
  f.setProtocol('ethereum')

  const m = f.getMetadataMap()
  m.set('stack', `${err.stack}`)
  m.set('message', `${err.message}`)
  m.set('name', `${err.name}`)

  return f
}

export class NetworkError extends Error {
  constructor(e: unknown, name?: string) {
    super()

    if (name !== undefined) {
      this.name = name
    }

    if (e instanceof Error) {
      this.stack = e.stack
      this.message = e.message
      this.cause = e.cause
    } else {
      this.message = `${e}`
    }
  }
}
