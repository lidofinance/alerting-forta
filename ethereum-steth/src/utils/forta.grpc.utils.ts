import { formatAddress, isZeroAddress } from 'forta-agent/dist/cli/utils'
import * as agent_pb from '../proto/agent_pb'
import { BlockEvent, TransactionEvent } from '../proto/agent_pb'
import * as alert_pb from '../proto/alert_pb'
import { BlockEvent as FortaBlockEvent, Finding, TransactionEvent as FortaTxEvent } from 'forta-agent'
import { EventType } from 'forta-agent/dist/sdk/event.type'
import { getContractAddress } from 'ethers/lib/utils'

export function createBlockEventFromGrpcRequest(request: agent_pb.EvaluateBlockRequest): FortaBlockEvent {
  const blockEvent = <agent_pb.BlockEvent>request.getEvent()
  const block: agent_pb.BlockEvent.EthBlock = <agent_pb.BlockEvent.EthBlock>blockEvent.getBlock()

  const blok = {
    difficulty: block.getDifficulty(),
    extraData: block.getExtradata(),
    gasLimit: block.getGaslimit(),
    gasUsed: block.getGasused(),
    hash: block.getHash(),
    logsBloom: block.getLogsbloom(),
    miner: formatAddress(block.getMiner()),
    mixHash: block.getMixhash(),
    nonce: block.getNonce(),
    number: parseInt(block.getNumber()),
    parentHash: block.getParenthash(),
    receiptsRoot: block.getReceiptsroot(),
    sha3Uncles: block.getSha3uncles(),
    size: block.getSize(),
    stateRoot: block.getStateroot(),
    timestamp: parseInt(block.getTimestamp()),
    totalDifficulty: block.getTotaldifficulty(),
    transactions: block.getTransactionsList(),
    transactionsRoot: block.getTransactionsroot(),
    uncles: block.getUnclesList(),
  }

  const network = <agent_pb.BlockEvent.Network>blockEvent.getNetwork()

  return new FortaBlockEvent(blockEvent.getType(), parseInt(network.getChainid()), blok)
}

export function fortaFindingToGrpc(finding: Finding): alert_pb.Finding {
  const out: alert_pb.Finding = new alert_pb.Finding()

  out.setProtocol(finding.protocol)
  out.setSeverity(Number(finding.severity))
  out.setType(Number(finding.type))
  out.setAlertid(finding.alertId)
  out.setName(finding.name)
  out.setDescription(finding.description)
  out.setPrivate(false)
  out.setAddressesList(finding.addresses)
  out.setUniquekey(finding.uniqueKey)
  out.setTimestamp(finding.timestamp.toString())

  const metadata = out.getMetadataMap()
  for (const key in finding.metadata) {
    metadata.set(key, finding.metadata[key])
  }

  return out
}

export function createTransactionEventFromGrpcRequest(request: agent_pb.EvaluateTxRequest): FortaTxEvent {
  const txEvent = <agent_pb.TransactionEvent>request.getEvent()
  const type: EventType = txEvent.getType()
  const network = <BlockEvent.Network>txEvent.getNetwork()
  const transaction = <TransactionEvent.EthTransaction>txEvent.getTransaction()
  const logList = <Array<TransactionEvent.Log>>txEvent.getLogsList()
  const block = <TransactionEvent.EthBlock>txEvent.getBlock()
  const tracesList = txEvent.getTracesList()

  const tx = {
    hash: transaction.getHash(),
    from: formatAddress(transaction.getFrom()),
    to: transaction.getTo() ? formatAddress(transaction.getTo()) : null,
    nonce: parseInt(transaction.getNonce()),
    gas: transaction.getGas(),
    gasPrice: transaction.getGasprice(),
    value: transaction.getValue(),
    data: transaction.getInput(),
    r: transaction.getR(),
    s: transaction.getS(),
    v: transaction.getV(),
  }
  const addresses = {
    [tx.from]: true,
  }
  if (tx.to) {
    addresses[tx.to] = true
  }

  const traces = []
  for (const trace of tracesList) {
    const action = <TransactionEvent.TraceAction>trace.getAction()
    const result = trace.getResult()

    traces.push({
      action: {
        callType: action.getCalltype(),
        to: formatAddress(action.getTo()),
        input: action.getInput(),
        from: formatAddress(action.getFrom()),
        value: action.getValue(),
        init: action.getInit(),
        address: formatAddress(action.getAddress()),
        balance: action.getBalance(),
        refundAddress: formatAddress(action.getRefundaddress()),
      },
      blockHash: trace.getBlockhash(),
      blockNumber: trace.getBlocknumber(),
      result: {
        gasUsed: result !== undefined ? result.getGasused() : '',
        address: result !== undefined ? result.getAddress() : '',
        code: result !== undefined ? result.getCode() : '',
        output: result !== undefined ? result.getOutput() : '',
      },
      subtraces: trace.getSubtraces(),
      traceAddress: trace.getTraceaddressList(),
      transactionHash: trace.getTransactionhash(),
      transactionPosition: trace.getTransactionposition(),
      type: trace.getType(),
      error: trace.getError(),
    })
  }

  const logs = logList.map((log) => {
    const address = formatAddress(log.getAddress())
    addresses[address] = true

    return {
      address: address,
      topics: log.getTopicsList(),
      data: log.getData(),
      logIndex: parseInt(log.getLogindex()),
      blockNumber: parseInt(log.getBlocknumber()),
      blockHash: log.getBlockhash(),
      transactionIndex: parseInt(log.getTransactionindex()),
      transactionHash: log.getTransactionhash(),
      removed: log.getRemoved(),
    }
  })

  let contractAddress = null
  if (isZeroAddress(transaction.getTo())) {
    contractAddress = formatAddress(getContractAddress({ from: transaction.getFrom(), nonce: transaction.getNonce() }))
  }

  return new FortaTxEvent(
    type,
    parseInt(network.getChainid()),
    tx,
    traces,
    addresses,
    {
      hash: block.getBlockhash(),
      number: parseInt(block.getBlocknumber()),
      timestamp: parseInt(block.getBlocknumber()),
    },
    logs,
    contractAddress,
  )
}
