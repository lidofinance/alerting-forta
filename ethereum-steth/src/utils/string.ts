import { ETH_DECIMALS } from './constants'
import BigNumber from 'bignumber.js'

export function etherscanNft(address: string, id: number | string): string {
  const subpath = process.env.FORTA_AGENT_RUN_TEAR == 'testnet' ? 'goerli.' : ''
  return `[${id}](https://${subpath}etherscan.io/nft/${address}/${id})`
}

export function etherscanAddress(address: string): string {
  const subpath = process.env.FORTA_AGENT_RUN_TIER == 'testnet' ? 'goerli.' : ''
  return `[${address}](https://${subpath}etherscan.io/address/${address})`
}

export function toEthString(wei: BigNumber): string {
  return wei.dividedBy(ETH_DECIMALS).toFixed(3) + ' ETH'
}
