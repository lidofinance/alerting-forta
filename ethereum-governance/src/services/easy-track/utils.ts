import { SAFES } from 'constants/easy-track'
import { STONKS } from 'constants/stonks'
import { getSafeAddress, etherscanAddress } from '../../shared/string'
import { Blockchain } from '../../shared/contracts'

export const getMotionCreatorNamedLink = (address: string) => {
  const ethSafes = SAFES.Ethereum
  const safeName = ethSafes.find((safeData: string[]) => safeData[0].toLowerCase() === address.toLowerCase())
  if (safeName) {
    return getSafeAddress(address, safeName[1])
  } else {
    return etherscanAddress(address)
  }
}

export const getStonksContractInfo = (address: string) => {
  return STONKS.find((c) => c.address === address)
}

export const getSafeNameByAddress = (address: string) => {
  let safeName = address
  Object.keys(SAFES).forEach((key) => {
    const safe = SAFES[key as Blockchain].find(
      (safeData: string[]) => safeData[0].toLowerCase() === address.toLowerCase(),
    )
    if (safe) {
      safeName = safe[1]
      return
    }
  })
  return safeName
}
