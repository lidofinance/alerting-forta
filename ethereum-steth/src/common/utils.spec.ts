import { Finding, FindingSeverity, FindingType } from "forta-agent";
import { mergeFindings } from "./utils";

describe("utils", () => {
  let timeSpy: jest.SpyInstance;

  beforeAll(() => {
    timeSpy = jest.spyOn(Date, "now");
    timeSpy.mockImplementation(() => new Date("2023-12-31"));
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it("should merge findings", () => {
    const ids = ["SOME-CUTE-ALERT", "SOME-FAKE-ALERT", "SOME-SOSO-ALERT"];
    const wordsList = ["🌷🌷🌷", "🍯🍯🍯", "🌸🌸🌸"];

    const findings = Array.from({ length: 51 }, (_, i) => {
      return Finding.fromObject({
        name: "🚨 Alert",
        description: wordsList[i % wordsList.length],
        alertId: ids[i % ids.length],
        severity: FindingSeverity.Low,
        type: FindingType.Degraded,
      });
    });

    const reducedFindings = mergeFindings(findings);
    expect(reducedFindings).toMatchSnapshot();
    expect(reducedFindings.length).toBe(3);
  });
});
