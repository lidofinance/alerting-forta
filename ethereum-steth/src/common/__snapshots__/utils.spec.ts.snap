// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils should merge findings 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "SOME-CUTE-ALERT",
    "description": "🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷
---
🌷🌷🌷",
    "labels": [],
    "metadata": {
      "args": ",,,,,,,,,,,,,,,,",
    },
    "name": "🚨 Alert",
    "protocol": "ethereum",
    "severity": 2,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 3,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "SOME-FAKE-ALERT",
    "description": "🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯
---
🍯🍯🍯",
    "labels": [],
    "metadata": {
      "args": ",,,,,,,,,,,,,,,,",
    },
    "name": "🚨 Alert",
    "protocol": "ethereum",
    "severity": 2,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 3,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "SOME-SOSO-ALERT",
    "description": "🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸
---
🌸🌸🌸",
    "labels": [],
    "metadata": {
      "args": ",,,,,,,,,,,,,,,,",
    },
    "name": "🚨 Alert",
    "protocol": "ethereum",
    "severity": 2,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 3,
    "uniqueKey": "",
  },
]
`;
